File: penna/src/fishing_pop.cpp
#include "fishing_pop.hpp"

// Returns true if animal gets fished.
struct FishingPredicate{
	using age_t = unsigned int;
	double f_rate;
	age_t f_age;
	FishingPredicate( double const & rate, 
					  age_t const & age ) : f_rate(rate), f_age(age) {};
	bool operator()(Penna::Animal const & animal){
		return (animal.age() > f_age) and (drand48() < f_rate);
	}
};

namespace Penna {

void FishingPopulation::change_fishing(	double const & f_rate, 
										age_t const & f_age){
	fishing_rate_ = f_rate;
	fishing_age_  = f_age;
}

void FishingPopulation::step(){
	Population::step();
	pop_.remove_if( FishingPredicate(fishing_rate_, fishing_age_) );
}

};

File: penna/pypenna/src/fishing_pop.py
from src.population import Population
import random


def FishingPrediacte(animal, f_rate, f_age):
    return animal.age() > f_age and f_rate > random.random()


class FishingPopulation(Population):
    def __init__(self, how_many, nmax=10000):
        super().__init__(how_many, nmax)
        self._fishing_rate = 0
        self._fishing_age = 0

    def change_fishing(self, n_frate, n_fage):
        self._fishing_rate = n_frate
        self._fishing_age = n_fage

    def step(self):
        super().step()
        self._pop[:] = [animal for animal in self._pop 
                        if not FishingPrediacte(animal, 
                            self._fishing_rate, self._fishing_age)]